dump/plugins/route.py

```diff
@@ -0,0 +1,296 @@
+        return self.ret_temp
+
+    def add_to_ret_template(self, table, db, keys, err, add_to_tables_not_found=True):
+        if not err and keys:
+            self.ret_temp[db]["keys"].extend(keys)
+            return keys
+        elif add_to_tables_not_found:
+            self.ret_temp[db]["tables_not_found"].extend([table])
+        return []
+
+    def init_route_config_info(self):
+        req = MatchRequest(db="CONFIG_DB", table="STATIC_ROUTE", key_pattern=self.dest_net, ns=self.ns)
+        ret = self.match_engine.fetch(req)
+        return self.add_to_ret_template(req.table, req.db, ret["keys"], ret["error"])
+
+    def init_route_appl_info(self):
+        req = MatchRequest(db="APPL_DB", table="ROUTE_TABLE", key_pattern=self.dest_net, ns=self.ns)
+        ret = self.match_engine.fetch(req)
+        self.add_to_ret_template(req.table, req.db, ret["keys"], ret["error"])
+
+    def init_asic_route_entry_info(self):
+        nh_id_field = "SAI_ROUTE_ENTRY_ATTR_NEXT_HOP_ID"
+        req = MatchRequest(db="ASIC_DB", table="ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY", key_pattern=get_route_pattern(self.dest_net),
+                           ns=self.ns, return_fields=[nh_id_field])
+        ret = self.match_engine.fetch(req)
+        keys = self.add_to_ret_template(req.table, req.db, ret["keys"], ret["error"])
+        asic_route_entry = keys[0] if keys else ""
+        vr = get_vr_oid(asic_route_entry)
+        nh_id = ret["return_values"].get(asic_route_entry, {}).get(nh_id_field, "")
+        return nh_id, vr
+
+    def init_asic_vr_info(self, vr):
+        ret = {}
+        if vr:
+            req = MatchRequest(db="ASIC_DB", table="ASIC_STATE:SAI_OBJECT_TYPE_VIRTUAL_ROUTER", key_pattern=vr, ns=self.ns)
+            ret = self.nhgrp_match_engine.fetch(req)
+        self.add_to_ret_template("ASIC_STATE:SAI_OBJECT_TYPE_VIRTUAL_ROUTER", "ASIC_DB", ret.get("keys", []), ret.get("error", ""))
+
+    def init_asic_nh(self):
+        self.nh_type = self.get_nh_type()
+        nh_ex = NHExtractor.initialize(self)
+        nh_ex.collect()
+
+    def get_nh_type(self):
+        if not self.nh_id:
+            return "DROP"
+        oid = self.nh_id.split(":")[-1]
+        for nh_type in [NH_GRP, NH, RIF, CPU_PORT]:
+            if oid.startswith(OID_HEADERS.get(nh_type, "")):
+                return nh_type
+        return "DROP"
+
+
+class NHExtractor(object):
+    """
+    Base Class for NH_ID Type
+    """
+    @staticmethod
+    def initialize(route_obj):
+        if route_obj.nh_type == NH:
+            return SingleNextHop(route_obj)
+        elif route_obj.nh_type == NH_GRP:
+            return MultipleNextHop(route_obj)
+        elif route_obj.nh_type == RIF:
+            return DirecAttachedRt(route_obj)
+        elif route_obj.nh_type == CPU_PORT:
+            return CPUPort(route_obj)
+        return NHExtractor(route_obj)
+
+    def __init__(self, route_obj):
+        self.rt = route_obj
+
+    def collect(self):
```

Comment:
Looks like it is abstract method which is required to be overridden by derived classes.
At least I as I see all derived classes always implement it so it looks likes it is must.
If it is the case then there is recommended exception to raise in base class in such abstract methods: ```NotImplementedError```

Reply:
This is invoked from here:
```
def init_asic_nh(self):
        self.nh_type = self.get_nh_type()
        nh_ex = NHExtractor.initialize(self) # This will call the actual derived class of NHExtractor
        nh_ex.collect()
```

```
@staticmethod
    def initialize(route_obj):
        if route_obj.nh_type == NH:
            return SingleNextHop(route_obj)
        elif route_obj.nh_type == NH_GRP:
            return MultipleNextHop(route_obj)
        elif route_obj.nh_type == RIF:
            return DirecAttachedRt(route_obj)
        elif route_obj.nh_type == CPU_PORT:
            return CPUPort(route_obj)
        return NHExtractor(route_obj)
```

When a nh_type apart from NH, NH_GRP, RIF or CPU_PORT is provided, the base class itself is initialized, so a NotImplemented  exception will cause the dump module to not exit gracefully.